File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/map/NLayerGroups.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.map

import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asList

internal data class NLayerGroups(val groups: List<String>) {

    fun useWithEnableAndDisableGroups(func: (enableGroups: List<String>, disableGroups: List<String>) -> Unit) {
        val disableGroups = ALL_LAYER_GROUPS.groups - groups.toSet()
        func(groups, disableGroups)
    }

    companion object {
        fun fromRawList(rawList: Any): NLayerGroups = NLayerGroups(rawList.asList { it.toString() })

        private const val building: String = "building"
        private const val traffic: String = "ctt"
        private const val transit: String = "transit"
        private const val bicycle: String = "bike"
        private const val mountain: String = "mountain"
        private const val cadastral: String = "landparcel"

        private val ALL_LAYER_GROUPS: NLayerGroups = NLayerGroups(
            listOf(building, traffic, transit, bicycle, mountain, cadastral)
        )
    }
}

File: ios/Classes/model/map/NLayerGroups.swift
import NMapsMap

internal struct NLayerGroups {
    let groups: Set<String>

    func useWithEnableAndDisableGroups(fun: (_ enableGroups: Set<String>, _ disableGroups: Set<String>) -> Void) {
        let disableGroups = NLayerGroups.allLayerGroups.groups.subtracting(groups)

        fun(groups, disableGroups)
    }

    static func fromRawArr(rawArr: Any) -> NLayerGroups {
        NLayerGroups(groups: Set(asArr(rawArr, elementCaster: asString)))
    }

    private static let building: String = "building"
    private static let traffic: String = "ctt"
    private static let transit: String = "transit"
    private static let bicycle: String = "bike"
    private static let mountain: String = "mountain"
    private static let cadastral: String = "landparcel"

    private static let allLayerGroups: NLayerGroups = NLayerGroups(groups: [
        building, traffic, transit, bicycle, mountain, cadastral
    ])
}